Change shell execution methods to be async.

cgranade opened the pull request:
With #43, we added the ability to provide custom routing, useful for implementing new kinds of shell messages. This turns out to have introduced some subtle deadlocking issues when trying to receive shell messages during the processing of another shell message. To resolve that, this PR makes the _breaking change_ of moving shell and iopub message handling to be entirely async.

This change is especially useful for implementing communication between the client and the kernel during execution requests (e.g.: interactive widgets).

vxfield commented on src/Servers/ShellServer.cs:
```diff
@@ -140,7 +143,7 @@ private void EventLoop(NetMQSocket socket)
                     // Get a service that can handle the message type and
                     // dispatch.
-                    router.Handle(nextMessage);
+                    Task.Run(() => router.HandleAsync(nextMessage));
```
Is it really ok to run all messages asynchronously?
Maybe some messages need to be processed in order by their corresponding router/handler.
So it might be better to have the router to deal with that, maybe just having a method to accept messages (StartHandleMessage), and the router decides whether to create new tasks for and run them async or not.
The same logic could be used for the handler, so the router does not make that decision, but the handler does.
This way we could give more control to the handler implementation on how to process messages (concurrently or not).
From the ShellServer and Router perspective, they are just starting things and they don't wait for the message to be processed, only for the message to be accepted.

cgranade replied:
The trouble is that even when a message needs to be handled synchronously with respect to other messages of the same kind, messages of other message types still need to be processed asynchronously. We've not done that correctly in previous versions, but as we increase the amount of client communication, that becomes more of an issue.

cgranade replied:
I've modified to using a queue, and to having handlers return either `null` to indicate no asynchronous completion is required, or a task that indicates that the remainder of handling can be performed asynchronously. This allows for ensuring messages are always properly ordered, but that processing can be nonblocking for long-running handlers (e.g.: execution).

vxfield commented on src/ShellRouting/IShellRouter.cs:
```diff
@@ -26,6 +27,16 @@ public interface IShellHandler
         ///     The incoming message to be handled.
         /// </param>
         public void Handle(Message message);
```
Why not to just have an async version of the message handling?
Less execution branches to maintain and test...

cgranade replied:
Sounds good, will remove this, especially since it's a breaking change already.

vxfield commented on src/ShellRouting/IShellRouter.cs:
```diff
@@ -26,6 +27,16 @@ public interface IShellHandler
+        /// <param name="message">
+        ///     The incoming message to be handled.
+        /// </param>
+        public async Task HandleAsync(Message message) =>
```
Perhaps have a generic return type like Task<MessageResponse<TResponse>>
where the MessageResponse contains some basic properties like Status (OK/Error) and a TResponse type that is the message response to the client.
If the router finds that the message response is not null, it automatically sends it back to the client.

In this case, the router needs to await for the response, and therefore handle the request queue.

cgranade replied:
I went with a slightly different approach, but that might be a good improvement in the future!

vxfield commented on src/ShellRouting/IShellRouter.cs:
```diff
@@ -58,7 +69,41 @@ public interface IShellRouter
         ///     ]]></code>
         /// </example>
-        public void RegisterHandler(string messageType, Action<Message> handler);
+        public void RegisterHandler(string messageType, Action<Message> handler)
```
as before, I would only keep the async version and "force" all handler implementers to "think asynchronously" :-)

cgranade commented on src/Engines/BaseEngine.cs:
```diff
@@ -460,11 +469,71 @@ public virtual void OnExecuteRequest(Message message)
                     }.AsReplyTo(message)
                 );
 
+                Task<ExecutionResult>? previousTask;
```
Good point, I'll refactor to make this a queue. As for client-vs-server, we don't directly have control over the client (in some cases, they may even be multiple clients, though never very many), such that we have to respect the current client behavior. In this case, current clients treat busy as a UI indicator, but still allow users to enqueue additional execution requests.

cgranade commented:
Thanks for the review, @vxfield! I've addressed your comments and changed to allow handlers to guarantee that handling starts synchronously, and can then proceed asynchronously, ensuring that handlers can require a strict ordering to messages while still not blocking the shell server thread.

vxfield commented on src/Servers/ShellServer.cs:
```diff
@@ -140,7 +143,11 @@ private void EventLoop(NetMQSocket socket)
+
+                    // If the handler requires additional processing, delegate
+                    // that to a background thread.
+                    if (task != null) Task.Run(() => task);
```
also, every time we do a Task, we completely lose the control on when that task will run, so there is no guarantee for order anymore

vxfield commented on src/Engines/BaseEngine.cs:
```diff
@@ -139,6 +141,8 @@ public ExecutedEventArgs(ISymbol symbol, ExecutionResult result, TimeSpan durati
         private Dictionary<string, Stack<IResultEncoder>> serializers = new Dictionary<string, Stack<IResultEncoder>>();
         private List<ISymbolResolver> resolvers = new List<ISymbolResolver>();
 
+        private readonly Stack<Lazy<Task<ExecutionResult?>>> executionQueue = new Stack<Lazy<Task<ExecutionResult?>>>();
```
I don't quite understand why we have an stack (last in, first out) that we are calling and using as a queue (first in, first out).
I guess the idea was to stack the task results so we can wait for them in the reverse order, but I think this was getting too complicated...

I wrote an example of how I would do it here:
https://github.com/vxfield/AsyncMessageHandler

You could continue to have the handler to be just an Action, or you could make the handler to be a subclass of an abstract MessageHandler class and the protected execute method.
Having it as a separate class might be better to encapsulate the handlers to their own "component" and their could set their individual message handling mode (see my example).

cgranade replied:
Thanks for your feedback! I've added a separate class to represent async handlers that need strict ordering of their messages, following your example. I think that cleaned up the code quite a bit.  Please let me know if you think there's anything else I should add. Thanks!

vxfield commented:
I just updated my example with Unit Testing

vxfield commented on src/KernelApplication.cs:
```diff
@@ -612,19 +612,28 @@ public virtual int StartKernel(ServiceProvider serviceProvider)
+            logger.LogDebug("Starting shell service.");
             shellServer.Start();
 
+            scope.Dispose();
```
Perhaps instead of explicitly calling dispose, wrap the code in an "using" statement?

cgranade replied:
I definitely considered a `using` block here, but it wound up making the code less readable in this case. Thanks for the suggestion, though!
